### PR TITLE
Add access to `full_path` index of uploaded files

### PR DESCRIPTION
**Pull request by JamminCoder**
**Description**
Supersedes #7538

Added `UploadedFile::getClientPath()` method.  
This returns the `full_path` index of uploaded files (when uploaded via directory).  
This feature is only available in PHP 8.1+. If the method is called below that version then `null` is returned.  

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by JamminCoder** [comment]
Ok, let's try again. I fixed all the linting errors, updated the style of my new code, so it should pass the workflow checks. 
I did not get to update the user guide and add documentation for the new method, but can do that shortly.

**Comment by JamminCoder** [comment]
Hmm, ok. I guess there are still errors. I'll figure those out.

**Comment by kenjis** [comment]
We still support PHP 7.4. So code must work on PHP 7.4.

File: system/HTTP/Files/UploadedFile.php [hunk 1]
```diff
@@ -75,13 +82,15 @@ class UploadedFile extends File implements UploadedFileInterface
      *
      * @param string $path         The temporary location of the uploaded file.
      * @param string $originalName The client-provided filename.
+     * @param string $clientPath   The webkit relative path of the uploaded file.
      * @param string $mimeType     The type of file as provided by PHP
      * @param int    $size         The size of the file, in bytes
      * @param int    $error        The error constant of the upload (one of PHP's UPLOADERRXXX constants)
      */
-    public function __construct(string $path, string $originalName, ?string $mimeType = null, ?int $size = null, ?int $error = null)
+    public function __construct(string $path, string $originalName, ?string $clientPath, ?string $mimeType = null, ?int $size = null, ?int $error = null)
```
Comment (samsonasik) [inline system/HTTP/Files/UploadedFile.php:90]: add param in the middle is BC break.

Comment (JamminCoder) [reply system/HTTP/Files/UploadedFile.php:90]: Ah, yeah. Perhaps I should move the param to the end for now, so it doesn't interfere with any instances where I didn't update the arguments.

Comment (kenjis) [reply system/HTTP/Files/UploadedFile.php:90]: Yes, this must be added in the last.

Comment (JamminCoder) [reply system/HTTP/Files/UploadedFile.php:90]: And I forgot to allow that param as null

File: system/HTTP/Files/UploadedFile.php [hunk 2]
```diff
@@ -267,6 +276,15 @@ public function getClientName(): string
         return $this->originalName;
     }
 
+    /**
+     * (PHP 8.1+)
+     * Returns the webkit relative path of the uploaded file on directory uploads.
+     */
+    public function getClientPath(): string|null
```
Comment (samsonasik) [inline system/HTTP/Files/UploadedFile.php:283]: use nullable instead

```suggestion
    public function getClientPath(): ?string
```

File: tests/system/HTTP/Files/FileCollectionTest.php [hunk 3]
```diff
@@ -38,11 +38,12 @@ public function testAllReturnsValidSingleFile()
...
-                'name'     => 'someFile.txt',
-                'type'     => 'text/plain',
-                'size'     => '124',
-                'tmp_name' => '/tmp/myTempFile.txt',
-                'error'    => 0,
+                'name'      => 'someFile.txt',
+                'type'      => 'text/plain',
+                'size'      => '124',
+                'tmp_name'  => '/tmp/myTempFile.txt',
+                'full_path' => 'tmp/myTempFile.txt',
```
Comment (kenjis) [inline tests/system/HTTP/Files/FileCollectionTest.php:45]: This full path is wrong because `name` is `someFile.txt`. Please make it more realistic.

Comment (JamminCoder) [reply tests/system/HTTP/Files/FileCollectionTest.php:45]: Yes. Good point.

File: system/HTTP/Files/UploadedFile.php [hunk 4]
```diff
@@ -75,13 +82,15 @@ class UploadedFile extends File implements UploadedFileInterface
      *
      * @param string $path         The temporary location of the uploaded file.
      * @param string $originalName The client-provided filename.
+     * @param string $clientPath   The webkit relative path of the uploaded file.
```
Comment (kenjis) [inline system/HTTP/Files/UploadedFile.php:85]: Move this line after the `$error` line.

File: tests/system/HTTP/Files/FileCollectionTest.php [hunk 5]
```diff
@@ -54,6 +55,7 @@ public function testAllReturnsValidSingleFile()
         $this->assertInstanceOf(UploadedFile::class, $file);
 
         $this->assertSame('someFile.txt', $file->getName());
+        $this->assertSame('tmp/myTempFile.txt', $file->getClientPath());
```
Comment (kenjis) [inline tests/system/HTTP/Files/FileCollectionTest.php:58]: Add a test case for null.

File: tests/system/HTTP/Files/FileCollectionTest.php [hunk 6]
```diff
@@ -55,6 +56,12 @@ public function testAllReturnsValidSingleFile()
 
         $this->assertSame('someFile.txt', $file->getName());
         $this->assertSame(124, $file->getSize());
+
+        if (version_compare(PHP_VERSION, '8.1', '>=')) {
```
Comment (kenjis) [inline tests/system/HTTP/Files/FileCollectionTest.php:60]: Because we cannot upload in testing and `$_FILES` is given, we cannot write test like this.
Just add another test with `$_FILES` without `full_path` is enough.

File: user_guide_src/source/libraries/uploaded_files.rst [hunk 7]
```diff
@@ -304,6 +304,14 @@ version, use ``getMimeType()`` instead:
 
 .. literalinclude:: uploaded_files/015.php
 
+getClientPath()
+---------------
+
```
Comment (kenjis) [inline user_guide_src/source/libraries/uploaded_files.rst:309]:
```suggestion

.. versionadded:: 4.4.0

```

**Comment by kenjis** [comment]
Add a short description in the changelog to let devs know this enhancement.
https://github.com/codeigniter4/CodeIgniter4/blob/4.4/user_guide_src/source/changelogs/v4.4.0.rst#libraries

File: tests/system/HTTP/Files/FileCollectionTest.php (hunk 6 above)
Comment (JamminCoder) [reply tests/system/HTTP/Files/FileCollectionTest.php]: Yes. Just updated that. I realized how stupid that was.

**Comment by JamminCoder** [comment]
The 1 failed check is this:
image
It doesn't look like any of the changes I made would cause it, though I can look into fixing this one too.

**Comment by kenjis** [comment]
It is because of PHPStan update.
I sent #7542 to `develop` already.

**Review by kenjis — APPROVED** [review]
LGTM! Thank you.

File: system/HTTP/Files/UploadedFile.php [hunk 8]
```diff
@@ -78,10 +85,12 @@ class UploadedFile extends File implements UploadedFileInterface
      * @param string $mimeType     The type of file as provided by PHP
      * @param int    $size         The size of the file, in bytes
      * @param int    $error        The error constant of the upload (one of PHP's UPLOADERRXXX constants)
+     * @param string $clientPath   The webkit relative path of the uploaded file.
      */
-    public function __construct(string $path, string $originalName, ?string $mimeType = null, ?int $size = null, ?int $error = null)
+    public function __construct(string $path, string $originalName, ?string $mimeType = null, ?int $size = null, ?int $error = null, ?string $clientPath = null)
     {
         $this->path             = $path;
+        $this->clientPath       = $clientPath;
```
Comment (samsonasik) [inline system/HTTP/Files/UploadedFile.php:93]: move this to last after line 98

**Review by kenjis — APPROVED** [review]

**Review by samsonasik — APPROVED** [review]
